File: Python/1079-total_sales_of_supply_chain.py
from collections import deque


def total_sales(graph, sink_nodes, unit_price, rate):
    if len(graph) == 1:
        return unit_price * graph[0][0]
    ans = 0.
    rate /= 100
    price = [0.] * len(graph)
    price[0] = unit_price
    queue = deque([0])
    while queue:
        u = queue.popleft()
        for v in graph[u]:
            price[v] = price[u] * (1 + rate)
            if v not in sink_nodes:
                queue.append(v)
            else:
                ans += price[v] * graph[v][0]
    return ans


if __name__ == '__main__':
    n_node, unit_price, rate = [float(i) for i in input().split()]
    n_node = int(n_node)
    graph = [None] * n_node
    sink_nodes = set()
    for i in range(n_node):
        count, *adj_nodes = [int(i) for i in input().split()]
        if count == 0:
            sink_nodes.add(i)
        graph[i] = adj_nodes
    print(f'{total_sales(graph, sink_nodes, unit_price, rate):.1f}')

File: C++/1079-total_sales_of_supply_chain.cc
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_set>
#include <iomanip>


using Graph = std::vector<std::vector<int>>;


double total_sales(const Graph &graph,
                   const std::unordered_set<int> &sink_nodes,
                   double unit_price, double rate)
{
    if (graph.size() == 1) {
        return unit_price * graph[0][0];
    } // only has one node
    double ans = 0.;
    rate /= 100;
    std::vector<double> price(graph.size());
    price[0] = unit_price;
    std::queue<int> queue;
    queue.push(0);
    while (!queue.empty()) {
        int u = queue.front();
        queue.pop();
        for (auto &v : graph[u]) {
            price[v] = price[u] * (1 + rate);
            if (sink_nodes.find(v) == sink_nodes.end()) {
                queue.push(v);
            } else {
                ans += price[v] * graph[v][0];
            }
        }
    }
    return ans;
}


int main() {

    std::ios::sync_with_stdio(false);

    int n_node;
    double unit_price, rate;
    std::cin >> n_node >> unit_price >> rate;
    Graph graph(n_node);
    std::unordered_set<int> sink_nodes;
    for (int i=0; i<n_node; ++i) {
        int count;
        std::cin >> count;
        if (count == 0) {
            sink_nodes.insert(i);
            ++count;
        }
        while (count--) {
            int v;
            std::cin >> v;
            graph[i].push_back(v);
        }
    }
    
    std::cout << std::fixed << std::setprecision(1)
              << total_sales(graph, sink_nodes, unit_price, rate) << '\n'
              << std::defaultfloat;

    return 0;
}
